build/lib/src/generate/expected_outputs.dart

```diff
@@ -42,59 +42,74 @@ extension on AssetId {
 }
 
 abstract class _ParsedBuildOutputs {
-  static final RegExp _captureGroup = RegExp('{{}}');
+  static final RegExp _captureGroup = RegExp('{{([\\w\\d]*)}}');
 
   _ParsedBuildOutputs();
 
   factory _ParsedBuildOutputs.parse(
       Builder builder, String input, List<String> outputs) {
-    final groups = _captureGroup.allMatches(input).iterator;
+    final groups = _captureGroup.allMatches(input);
 
-    if (!groups.moveNext()) {
+    if (groups.isEmpty) {
       // The input does not contain a capture group, so we should simply match
       // all assets whose paths ends with the desired input.
       return _SuffixBuildOutputs(input, outputs);
     }
 
-    final match = groups.current;
-
-    if (groups.moveNext()) {
-      throw ArgumentError(
-        'The builder `$builder` declares an input "$input" which contains '
-        'multiple groups ("{{}}"). This is not supported.',
-      );
+    final regexBuffer = StringBuffer();
+    var positionInInput = 0;
+    if (input.startsWith('^')) {
+      regexBuffer.write('^');
+      positionInInput = 1;
     }
 
-    // When using a capture group in the build input, it must also be used in
-    // every output to ensure outputs have unique names.
-    for (final output in outputs) {
-      if (_captureGroup.allMatches(output).length != 1) {
+    // Builders can have multiple capture groups, which are disambiguated by
+    // their name. Names can be empty as well: `{{}}` is a valid capture group.
+    final names = <String>{};
```

Comment:
I would actually just use a list here, and do a `contains` check below. There will never be a large number of things in it, so a list is actually more efficient. And then we don't need to convert it to a list later.

We also rely on the insertion order iteration, which not all sets do. If we do leave it as a set I would explicitly use the `LinkedHashSet` constructor (even though this is the default type), with a comment about insertion order iteration. Just so it is very explicit.

Reply:
I agree that being more explicit about insertion order is important, I've changed this to use a list.